xe.1
.Dd August 1, 2023
.Dt XE 1
.Os
.Sh NAME
.Nm xe
.Nd execute a command for every argument
.Sh SYNOPSIS
.Nm
.Op Fl 0FLRnqv
.Oo Fl I Ar replace-arg Oc
.Op Fl N Ar maxargs
.Op Fl j Ar maxjobs
.Ar command\ ...
.Nm
.Op Ar flags\ ...
.Fl p Ar pattern Ar command\ ...
.Oo Cm \&+ Ar pattern Ar command\ ... Oc Ns ...
.Nm
.Op Ar flags\ ...
.Fl f Ar argfile Ar command\ ...
.Nm
.Op Ar flags\ ...
.Fl s Ar shellscript
.Nm
.Op Ar flags\ ...
.Fl a Ar command\ ... Cm -- Ar args\ ...
.Nm
.Op Ar flags\ ...
.Fl A Ar argsep Ar command\ ... Ar argsep Ar args\ ...
.Sh DESCRIPTION
The
.Nm
utility constructs command lines from specified arguments,
combining some of the best features of
.Xr xargs 1
and
.Xr apply 1 .
.Pp
.Nm
means
.Dq execute for every ... .
.Pp
.Nm
supports different methods to specify arguments to commands:
.Bl -tag -width Ds
.It Ar command\ ...
By default, arguments - separated by newlines -
are read from the standard input.
The resulting command is constructed from the command line parameters,
replacing
.Ar replace-arg
with the read argument, and is executed with
.Xr execvp 3 .
.Pp
In this mode, no shell is involved and
.Ar replace-arg
must appear as a word on its own, i.e.
.Sq foo {} bar
will work, but
.Sq foo{} bar
will not, where {} is the default value for
.Ar replace-arg .
.Pp
If no argument is specified, the default is
.Sq Ic printf %s\en .
.It Fl f Ar argfile
Read arguments from
.Ar argfile ,
instead of the standard input.
.Pp
This does not close the standard input for execution,
it is passed to the forked process.
.It Fl s Ar shellscript
In this mode, the single parameter
.Ar shellscript
is executed using
.Ic sh -c .
In the script, the specified arguments can be accessed using $1, $2, ...
.Pp
For example:
.Dl echo \(aqa\enb\(aq | xe -N2 \-s \(aqecho $2 $1\(aq
.It Fl a Ar command\ ... Cm -- Ar args\ ...
In this mode, everything after
.Cm --
is passed as
.Ar args
to
.Ar command .
.It Fl A Ar argsep Ar command\ ... Ar argsep Ar args\ ...
Same as
.Fl a ,
but the custom argument separator
.Ar argsep
is used to distinguish between
.Ar command
and its
.Ar args .
.El
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl 0
Input arguments are separated by NUL bytes (instead of newlines, which
is the default).
.It Fl F
Fatal:
stop and exit when a command execution fails.
.It Fl L
Run the resulting commands with line-buffered output;
lines from two jobs will not interleave.
When used twice,
or with
.Fl vv ,
also prefix each line with the number of the job
(see
.Sx ENVIRONMENT )
in such a manner that the output can be piped to
.Sq Li sort -snk1
to group it.
.It Fl R
Return with status 122 when no arguments have been specified
(instead of 0, the default).
.Nm
never executes a command when no arguments are specified.
.It Fl n
Dry run: don't run the resulting commands, just print them.
.It Fl q
Quiet mode:
redirect standard output and standard error of commands to /dev/null.
.It Fl v
Verbose: print commands to standard error before running them.
When used twice, also print job id and exit status for each command.
.It Fl p
Enable
.Xr make 1 Ns \&- Ns
style percent rules.
The first argument of
.Ar command\ ...
is regarded as a pattern,
see
.Sx PERCENT RULES
below.
Patterns without a slash (or
.Sq Li \&*\&* )
are matched against the basenames only.
.Pp
Multiple runs of patterns and commands are separated by
.Sq Li \&+ .
Only the first matching percent rule is executed;
in case no pattern matches, no command is run.
.It Fl I Ar replace-arg
Replace first occurrence of
.Ar replace-arg
(default:
.Cm {} )
in the resulting command with the argument(s).
Pass an empty
.Ar replace-arg
to disable the replace function.
Contrary to
.Xr xargs 1
this will expand into multiple arguments when needed.
.It Fl N Ar maxargs
Pass up to
.Ar maxargs
arguments to each command (default: 1).
.br
Using
.Fl N0
will pass as many arguments as possible.
.It Fl j Ar maxjobs
Run up to
.Ar maxjobs
processes concurrently.
Using
.Fl j0
will run as many processes as there are CPU cores running.
If
.Ar maxjobs
ends with an
.Sq Ic x ,
it is regarded as a multiplier of the number of running CPU cores
(rounded down, but using at least one core).
.El
.Sh PERCENT RULES
The percent rules of
.Nm
are similar to the globs
of
.Xr sh 1
or
.Xr fnmatch 3 :
.Sq Li \&?
matches a single character that is not
.Sq Li \&/ .
.Sq Li \&/
matches one or multiple
.Sq Li \&/
in the string.
.Sq Li \&*
matches zero or more characters, but never
.Sq Li \&/ .
.Sq Li \&*\&*
matches zero or more characters, including
.Sq Li \&/ .
Note that all of these also match leading dots in file names.
.Pp
.Sq Li \&{ Ns Va a Ns \&, Ns Va b Ns \&, Ns Va c Ns \&}
matches either
.Va a , b
or
.Va c .
.Sq Li \&[ Ns Va abc Ns \&]
matches one of the characters
.Va abc
(but never
.Sq Li \&/ ) .
.Sq Li \&[ Ns \&! Ns Va abc Ns \&]
matches all characters but
.Va abc .
Alternatively,
.Sq Li \&[ Ns \&^ Ns Va abc Ns \&]
can be used too.
.Sq Li \&[ Ns Va a Ns \&- Ns Va c Ns \&]
matches any character in the range between
.Va a
and
.Va c
inclusive.
In character ranges, characters can be escaped using a backslash.
.Pp
In the pattern, a single occurrence of
.Sq Li \&%
matches one or more characters,
and replaces the first occurrence of
.Sq Li \&%
with the matched string in the remaining arguments,
which are then used as the command to be executed.
.Sh ENVIRONMENT
The environment variable
.Ev ITER
is passed to the child process and incremented on each command execution.
.Sh EXIT STATUS
.Nm
follows the convention of GNU and OpenBSD xargs:
.Bl -tag -compact -width Ds
.It 0
on success
.It 123
if any invocation of the command exited with status 1 to 254.
.It 124
if the command exited with status 255
.It 125
if the command was killed by a signal
.It 126
if the command cannot be run
.It 127
if the command was not found
.It 1
if some other error occurred
.El
.Pp
Additionally, 122 is returned when
.Fl R
was passed and the command was never executed.
.Sh EXAMPLES
Compress all .c files in the current directory, using all CPU cores:
.Dl xe -a -j0 gzip -- *.c
Remove all empty files, using
.Xr lr 1 :
.Dl lr -U -t 'size == 0' | xe -N0 rm
Convert .mp3 to .ogg, using all CPU cores:
.Dl xe -a -j0 -s 'ffmpeg -i \&"${1}\&" \&"${1%.mp3}.ogg\&"' -- *.mp3
Same, using percent rules:
.Dl xe -a -j0 -p %.mp3 ffmpeg -i %.mp3 %.ogg -- *.mp3
Similar, but hiding output of ffmpeg, instead showing spawned jobs:
.Dl xe -ap -j0 -vvq '%.{m4a,ogg,opus}' ffmpeg -y -i {} out/%.mp3 -- *
.Sh SEE ALSO
.Xr apply 1 ,
.Xr parallel 1 ,
.Xr xapply 1 ,
.Xr xargs 1
.Sh AUTHORS
.An Leah Neukirchen Aq Mt leah@vuxu.org
.Sh LICENSE
.Nm
is in the public domain.
.Pp
To the extent possible under law,
the creator of this work
has waived all copyright and related or
neighboring rights to this work.
.Pp
.Lk http://creativecommons.org/publicdomain/zero/1.0/


NEWS.md
## HEAD

## 1.0 (2023-08-05)

* Slightly incompatible change: properly forward errors on exec,
  return 123 on any exit status 1-254 (like GNU xargs).
* Small bugfixes.

## 0.11 (2017-11-05)

* Add zsh completion.
* Fix `-F` to exit as soon as possible when a command failed.
* Fix when using `-s` and `-p` together.
* Better error handling for some rare situations.
* Made test suite more robust.

## 0.10 (2017-10-30)

* New flag `-p` to enable "percent rules" ala make(1), see the man
  page for details.
* New flag `-q` to run commands, hiding their output.
* Fix issues launching commands on FreeBSD and DragonFlyBSD.

## 0.9 (2017-07-17)

* New flag `-L` to enable line buffering on output.
* New flag `-LL` to prefix output lines with `$ITER`.
* `-j` now can scale with the available CPU cores
  (e.g. `-j0.5x` to use half the cores).
* `-vv` will trace `$ITER` and `exit status`.
* Fix issues related to having more children than we knew of.
* Documentation fixes and improvement by Larry Hynes.

## 0.8 (2017-06-02)

* Only read new arguments when we immediately need them

## 0.7 (2017-04-13)

* Error when the `-A` separator does not appear
* Bugfix: fix quoting of empty strings
* Documentation fixes


xe.c
/*
 * xe - simple xargs and apply replacement
 *
 * To the extent possible under law, Leah Neukirchen <leah@vuxu.org>
 * has waived all copyright and related or neighboring rights to this work.
 * http://creativecommons.org/publicdomain/zero/1.0/
 */

#include <sys/stat.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/wait.h>

#include <errno.h>
#include <fcntl.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

static char delim = '\n';
static char default_replace[] = "{}";
static char *replace = default_replace;
static char *argsep;
static char *fflag;
static char *sflag;

static int maxatonce = 1;
static int maxjobs = 1;
static int runjobs = 0;
static int failed = 0;
static int Aflag, Fflag, Lflag, Rflag, aflag, nflag, pflag, qflag, vflag;
static long iterations = 0;
static FILE *traceout;
static FILE *input;

static size_t argmax;

static char *buf;
static size_t buflen;
static size_t bufcap;

static char **args;
static size_t argslen;
static size_t argscap;
static size_t argsresv;

struct child {
	pid_t pid;  // 0 if empty slot
	long iter;
};
static struct child *children;

static char **inargs;

static char *line = 0;
static size_t linelen = 0;

static char *
getarg()
{
	if (aflag || Aflag) {
		if (inargs && *inargs)
			return *inargs++;
		else
			return 0;
	}

	int read = getdelim(&line, &linelen, delim, input);
	if (read == -1) {
		if (feof(input))
			return 0;
		else
			exit(1);
	}
	if (line[read-1] == delim)  // strip delimiter
		line[read-1] = 0;

	return line;
}

static int
mywait()
{
	int status;
	pid_t pid;

	pid = wait(&status);
	if (pid < 0)
		return !(errno == ECHILD);

	int i;
	for (i = 0; i < maxjobs; i++)
		if (children[i].pid == pid)
			goto my_child;

	return 1;

my_child:
	if (WIFEXITED(status)) {
		if (WEXITSTATUS(status) == 255) {
			fprintf(stderr, "xe: job %ld [pid %ld] exited with status 255\n", children[i].iter, (long)pid);
			exit(124);
		} else if (WEXITSTATUS(status) > 0) {
			if (Fflag) {
				fprintf(stderr, "xe: job %ld [pid %ld] exited with status %d\n", children[i].iter, (long)pid, WEXITSTATUS(status));
				exit(123);
			}
			failed = 1;
		}
	} else if (WIFSIGNALED(status)) {
		fprintf(stderr, "xe: job %ld [pid %ld] terminated by signal %d\n",
		    children[i].iter, (long)pid, WTERMSIG(status));
		exit(125);
	}

	if (vflag > 1) {
		fprintf(traceout, "%04ld< status %d\n",
		    children[i].iter, WEXITSTATUS(status));
		fflush(traceout);
	}

	children[i].pid = 0;
	runjobs--;

	return 1;
}

static void
sh_quote(const char *s)
{
	if (*s &&
	    !strpbrk(s, "\001\002\003\004\005\006\007\010"
	                "\011\012\013\014\015\016\017\020"
	                "\021\022\023\024\025\026\027\030"
	                "\031\032\033\034\035\036\037\040"
	                "`^#*[]=|\\?${}()'\"<>&;\177")) {
		fprintf(traceout, "%s", s);
		return;
	}

	fprintf(traceout, "\'");
	for (; *s; s++)
		if (*s == '\'')
			fprintf(traceout, "'\\''");
		else
			fprintf(traceout, "%c", *s);
	fprintf(traceout, "\'");
}

static int
trace()
{
	size_t i;

	for (i = 0; i < argslen; i++) {
		if (i > 0)
			fprintf(traceout, " ");
		sh_quote(args[i]);
	}
	fprintf(traceout, "\n");
	fflush(traceout);

	return 0;
}

static void
scanargs()
{
	char *s = buf;
	size_t i;

	for (i = 0; i < argslen; i++) {
		args[i] = s;
		s += strlen(s) + 1;
	}
	args[i] = 0;
}

static int
pusharg(const char *a)
{
	size_t l = strlen(a) + 1;   // including nul

	if (buflen >= argmax - l - argsresv || argslen + 1 >= argscap)
		return 0;

	if (buflen + l > bufcap) {
		while (buflen + l > bufcap)
			bufcap *= 2;
		buf = realloc(buf, bufcap);
		if (!args)
			exit(1);
	}

	memcpy(buf + buflen, a, l);
	buflen += l;
	argslen++;

	return 1;
}

static int
run()
{
	pid_t pid, lpid;

	while (runjobs >= maxjobs)
		mywait();
	runjobs++;
	iterations++;

	scanargs();

	if (!args[0]) {
		fprintf(stderr, "xe: no command\n");
		exit(126);
	}

	if (nflag) {
		trace();
		runjobs--;
		return 0;
	}

	int pipefd[2];
	if (Lflag) {
		if (pipe(pipefd) < 0)
			exit(126);
	}

	unsigned char status;
	int alivepipefd[2];
	if (pipe(alivepipefd) < 0)
		exit(126);
	fcntl(alivepipefd[0], F_SETFD, FD_CLOEXEC);
	fcntl(alivepipefd[1], F_SETFD, FD_CLOEXEC);

	pid = fork();
	if (pid == 0) {  // in child
		close(alivepipefd[0]);

		char iter[32];
		snprintf(iter, sizeof iter, "%ld", iterations);
		setenv("ITER", iter, 1);
		// redirect stdin to /dev/null when we read arguments from it
		// just close it if this fails
		if (input == stdin) {
			int fd = open("/dev/null", O_RDONLY);
			if (fd >= 0) {
				if (dup2(fd, 0) != 0)
					close(0);
				close(fd);
			} else {
				close(0);
			}
		}

		if (Lflag) {
			if (dup2(pipefd[1], 1) < 0)
				exit(126);
			close(pipefd[0]);
			close(pipefd[1]);
		}

		if (qflag) {
			int fd = open("/dev/null", O_RDONLY);
			if (fd >= 0) {
				dup2(fd, 2);
				dup2(fd, 1);
				close(fd);
			}
		}

		execvp(args[0], args);

		status = (errno == ENOENT ? 127 : 126);
		if (write(alivepipefd[1], &status, 1) != 1) {
			/* ignored */
		}
		fprintf(stderr, "xe: %s: %s\n", args[0], strerror(errno));
		_exit(status);
	} else if (pid < 0) {  // fork failed
		fprintf(stderr, "xe: %s: %s\n", args[0], strerror(errno));
		exit(126);
	}

	close(alivepipefd[1]);
	if (read(alivepipefd[0], &status, 1) == 1) {
		if (status == 126)
			exit(126);
		if (status == 127)
			exit(127);
	}
	close(alivepipefd[0]);

	if (Lflag) {
		long iter = iterations;
		lpid = fork();
		if (lpid == 0) {  // in line-logging child
			char *line = 0;
			size_t linelen = 0;

			close(0);
			close(pipefd[1]);
			FILE *input = fdopen(pipefd[0], "r");
			if (!input)
				exit(126);

			setvbuf(stdout, 0, _IOLBF, 0);

			while (1) {
				int rd = getdelim(&line, &linelen, '\n', input);
				if (rd == -1)
					exit(0);

				if (vflag > 1 || Lflag > 1)
					printf("%04ld= ", iter);
				fwrite(line, 1, rd, stdout);
			}
		}
		if (lpid < 0)
			exit(126);

		close(pipefd[0]);
		close(pipefd[1]);
	}

	if (vflag) {
		if (vflag > 1)
			fprintf(traceout, "%04ld> ", iterations);
		trace();
	}

	int i;
	for (i = 0; i < maxjobs; i++)
		if (!children[i].pid) {
			children[i].pid = pid;
			children[i].iter = iterations;
			break;
		}

	return 0;
}

void
toolong()
{
	fprintf(stderr, "xe: fixed argument list too long\n");
	exit(1);
}

int
parse_jobs(char *s)
{
	char *e;
	int n;

	if (*s && s[strlen(s) - 1] == 'x') {
#ifdef _SC_NPROCESSORS_ONLN
		n = (int)sysconf(_SC_NPROCESSORS_ONLN);
#else
		n = 1;
#endif
		double d = 0.0;
		errno = 0;
		d = strtod(s, &e);
		if (errno != 0 || *e != 'x' || d <= 0) {
			fprintf(stderr, "xe: can't parse multiplier '%s'.\n", s);
			exit(1);
		}
		n = (int)(d * n);
		if (n < 1)
			n = 1;
	} else if (strcmp(s, "j") == 0) {
		n = -1;
	} else {
		errno = 0;
		n = strtol(s, &e, 10);
		if (errno != 0 || *e) {
			fprintf(stderr, "xe: can't parse number '%s'.\n", s);
			exit(1);
		}
	}

#ifdef _SC_NPROCESSORS_ONLN
	if (n <= 0)
		n = (int)sysconf(_SC_NPROCESSORS_ONLN);
#endif
	if (n <= 0)
		n = 1;

	return n;
}

// perc is a variant of fnmatch, implemented here using left derivatives:
//
// differences to fnmatch:
// ** matches any substring
// % matches any nonempty substring (longest first), is kept for substitution
// [!...] and [^...] are supported
// [] matches nothing
// [...] never matches /
// {a,b,c} alternation
// {} matches empty string
// no special prefix . handling: *le matches .profile
// no collations in []

size_t perc_len;
char *perc_str;

char *
perc(char *pat, char *str, int lvl)
{
	char *s, *e;
	int neg, matched;
	size_t l;

	switch (*pat) {
	case '\0':
		if (lvl > 0) {
			fprintf(stderr, "xe: unmatched { in glob\n");
			exit(1);
		}
		return *str ? 0 : str;
	case '?':
		if (!*str || *str == '/')
			return 0;
		return perc(pat+1, str+1, lvl);
	case '/':
		if (*str != '/')
			return 0;
		while (*pat == '/')
			pat++;
		while (*str == '/')
			str++;
		return perc(pat, str, lvl);
	case '*':
		pat++;
		if (*pat == '*') {  // any substring
			while (*pat == '*')
				pat++;
			l = strlen(str) + 1;
		} else {  // any substring but no more than the next /
			if ((e = strchr(str, '/')))
				l = e - str + 1;
			else
				l = strlen(str) + 1;
		}
		while (l--)
			if ((s = perc(pat, str+l, lvl)))
				return s;
		return 0;
	case '%':
		// any nonempty substring
		for (l = strlen(str); l >= 1; l--)
			if ((s = perc(pat+1, str+l, lvl))) {
				perc_str = str;
				perc_len = l;
				return s;
			}
		return 0;
	case '[':
		if (!*str || *str == '/')
			return 0;
		pat++;
		neg = 0;
		if (*pat == '^' || *pat == '!') {
			pat++;
			neg = 1;
		}
		for (matched = 0; *pat && *pat != ']'; pat++) {
			if (*pat == '\\')
				pat++;
			if (pat[1] == '-' && pat[2] != ']') {
				if (pat[0] <= *str && *str <= pat[2])
					matched = 1;
				pat += 2;
			} else if (*str == *pat) {
				matched = 1;
			}
		}
		if (*pat != ']') {
			fprintf(stderr, "xe: unmatched [ in glob\n");
			exit(1);
		}
		return (neg ^ matched) ? perc(pat+1, str+1, lvl) : 0;
	case '{':
		e = 0;
		while (*pat++ != '}') {
			if (!e)
				e = perc(pat, str, lvl+1);
			// skip to next , or }
			for (l = 0;
			     *pat && !(l == 0 && (*pat == ',' || *pat == '}'));
			     pat++)
				if (*pat == '{')
					l++;
				else if (*pat == '}')
					l--;
				else if (*pat == '[')
					while (*pat && *pat != ']') {
						if (*pat == '\\' && pat[1])
							pat++;
						pat++;
					}
		}
		return e ? perc(pat, e, lvl) : 0;
	case ',':
	case '}':
		if (lvl > 0)
			return str;
		/* FALL THROUGH */
	default:
		return (*pat == *str) ? perc(pat+1, str+1, lvl) : 0;
	}
}

int
perc_match(char *pat, char *arg)
{
	if (!strchr(pat, '/') && !strstr(pat, "**")) {
		char *d = strrchr(arg, '/');
		if (d)
			arg = d + 1;
	}

	perc_len = 0;
	perc_str = 0;

	return perc(pat, arg, 0) != 0;
}

char *
perc_subst(char *pat, char *base, char *arg)
{
	static char buf[2048];
	size_t l;
	char *dir = base;

	if (strcmp(arg, replace) == 0)
		return base;

	if (!strchr(pat, '/') && !strstr(pat, "**")) {
		char *d = strrchr(base, '/');
		if (d)
			base = d + 1;
	}

	char *t = strchr(arg, '%');

	if (!t)
		return arg;

	if (perc_len)
		l = snprintf(buf, sizeof buf, "%.*s%.*s%.*s%.*s",
		    (int)(base - dir),
		    dir,

		    (int)(t - arg),
		    arg,

		    (int)perc_len,
		    perc_str,

		    (int)(arg + strlen(arg) - t),
		    t+1);
	else
		l = snprintf(buf, sizeof buf, "%.*s%.*s%s%.*s",
		    (int)(base - dir),
		    dir,

		    (int)(t - arg),
		    arg,

		    base,

		    (int)(arg + strlen(arg) - t),
		    t+1);

	if (l >= sizeof buf) {
		fprintf(stderr, "xe: result of percent substitution too long\n");
		exit(1);
	}

	return buf;
}

static void
trace_to_stderr()
{
	FILE *linestderr = fdopen(2, "w");
	if (linestderr) {
		setvbuf(linestderr, 0, _IOLBF, 0);
		traceout = linestderr;
	} else {
		traceout = stderr;
	}
}

int
main(int argc, char *argv[], char *envp[])
{
	int c, i, j, cmdend;
	char *arg;

	bufcap = 4096;
	buf = malloc(bufcap);

	argscap = 8192;
	args = malloc(sizeof args[0] * argscap);

	if (!buf || !args)
		exit(1);

	argmax = sysconf(_SC_ARG_MAX);
	while (*envp)  // subtract size of environment
		argmax -= strlen(*envp++) + 1 + sizeof *envp;
	argmax -= 4 * 1024;  // subtract 4k for safety
	if (argmax > 128 * 1024)  // upper bound
		argmax = 128 * 1024;
	if (argmax <= 0)  // lower bound
		argmax = _POSIX_ARG_MAX;

	traceout = stdout;

	while ((c = getopt(argc, argv, "+0A:FI:LN:Raf:j:npqs:v")) != -1)
		switch (c) {
		case '0': delim = '\0'; break;
		case 'A': argsep = optarg; Aflag++; break;
		case 'I': replace = optarg; break;
		case 'L': Lflag++; break;
		case 'N': maxatonce = atoi(optarg); break;
		case 'F': Fflag++; break;
		case 'R': Rflag++; break;
		case 'a': aflag++; break;
		case 'f': fflag = optarg; break;
		case 'j': maxjobs = parse_jobs(optarg); break;
		case 'n': nflag++; break;
		case 'p': pflag++; break;
		case 'q': qflag++; break;
		case 's': sflag = optarg; break;
		case 'v': vflag++; break;
		default:
			fprintf(stderr,
			    "Usage: %s [-0FLRnqv] [-I arg] [-N maxargs] [-j maxjobs] COMMAND...\n"
			    "     | -p PATTERN COMMAND... [+ PATTERN COMMAND...]...\n"
			    "     | -f ARGFILE COMMAND...\n"
			    "     | -s SHELLSCRIPT\n"
			    "     | -a COMMAND... -- ARGS...\n"
			    "     | -A ARGSEP COMMAND... ARGSEP ARGS...\n",
			    argv[0]);
			exit(1);
		}

	children = calloc(sizeof (struct child), maxjobs);
	if (!children)
		exit(1);

	if (!Lflag || vflag <= 1)
		trace_to_stderr();

	if (aflag || Aflag) {
		input = 0;
	} else if (!fflag || strcmp("-", fflag) == 0) {
		input = stdin;
	} else {
		input = fopen(fflag, "rb");
		if (!input) {
			fprintf(stderr, "xe: opening %s: %s\n",
			    fflag, strerror(errno));
			exit(1);
		}
		fcntl(fileno(input), F_SETFD, FD_CLOEXEC);
	}

	cmdend = argc;
	if (aflag) {  // find first -- in argv
		for (i = 1; i < argc; i++)
			if (strcmp(argv[i], "--") == 0) {
				inargs = argv + i+1;
				cmdend = i;
				break;
			}
		if (sflag) {  // e.g. on xe -s 'echo $1' -a 1 2 3
			cmdend = optind;
			inargs = argv + cmdend;
		}
	} else if (Aflag) {  // find first argsep after optind
		for (i = optind; i < argc; i++) {
			if (strcmp(argv[i], argsep) == 0) {
				inargs = argv + i+1;
				cmdend = i;
				break;
			}
		}
		if (i >= argc) {
			fprintf(stderr, "xe: '-A %s' used but no separator "
			    "'%s' found in command line.\n", argsep, argsep);
			exit(1);
		}
	}

	int keeparg = 0;

	if (pflag) {
		if (maxatonce != 1) {
			fprintf(stderr,
			    "xe: -p cannot be used together with -N.\n");
			exit(1);
		}

		while (1) {
			buflen = 0;
			argslen = 0;

			while (runjobs >= maxjobs)
				mywait();
			if (!(arg = getarg()))
				break;

			int n;
			for (n = optind, i = n + 1; n < cmdend; n = i + 1) {
				char *pat = argv[n];

				int matched = perc_match(pat, arg);

				if (sflag && matched) {
					pusharg("/bin/sh");
					pusharg("-c");
					pusharg(sflag);
					pusharg("/bin/sh");
				}

				for (i = n + 1; i < cmdend; i++) {
					if (argv[i][0] == '+' &&
					    argv[i][1] == '\0')
						break;

					if (matched &&
					    !pusharg(perc_subst(pat, arg, argv[i])))
						toolong();
				}

				if (matched) {
					run();
					break;
				}
			}
		}
	} else {
		while (1) {
			// check if there is an arg from a previous iteration
			if (!keeparg) {
				while (runjobs >= maxjobs)
					mywait();

				arg = getarg();
				if (!arg)
					break;
			}
			keeparg = 0;

			buflen = 0;
			argslen = 0;

			if (sflag) {
				pusharg("/bin/sh");
				pusharg("-c");
				pusharg(sflag);
				pusharg("/bin/sh");
			} else if (optind >= cmdend) {
				pusharg("printf");
				pusharg("%s\\n");
			}

			for (i = optind; i < cmdend; i++) {
				if (*replace && strcmp(argv[i], replace) == 0)
					break;
				if (!pusharg(argv[i]))
					toolong();
			}

			if (!pusharg(arg))
				toolong();
			i++;

			// reserve space for final arguments
			for (argsresv = 0, j = i; j < cmdend; j++)
				argsresv += 1 + strlen(argv[j]);

			// fill with up to maxatonce arguments
			for (j = 0; maxatonce < 1 || j < maxatonce-1; j++) {
				arg = getarg();
				if (!arg)
					break;
				if (!pusharg(arg)) {
					// we are out of space,
					// deal with this arg in the next iter
					keeparg = 1;
					break;
				}
			}

			for (argsresv = 0, j = i; j < cmdend; j++)
				if (!pusharg(argv[j]))
					toolong();

			run();
		}
	}

	while (mywait())
		;

	free(buf);
	free(args);
	free(line);

	if (Rflag && iterations == 0)
		return 122;
	if (failed)
		return 123;
	return 0;
}
